driver-core/src/main/com/mongodb/connection/ScramShaAuthenticator.java

```diff
@@ -0,0 +1,437 @@
+         * ServerKey       := HMAC(SaltedPassword, "Server Key")
+         * StoredKey       := H(ClientKey)
+         * ClientSignature := HMAC(StoredKey, AuthMessage)
+         * ClientProof     := ClientKey XOR ClientSignature
+         * ServerSignature := HMAC(ServerKey, AuthMessage)
+         */
+        String getClientProof(final String password, final String salt, final int iterationCount, final String authMessage)
+                throws SaslException {
+            String hashedPassword = encodeUTF8(h(decodeUTF8(password + salt)));
+            String cacheKey = encodeBase64(h(decodeUTF8(hashedPassword + ":" + salt + ":" + iterationCount)));
+
+            CachedKeys cachedKeys = KEY_CACHE.get(cacheKey);
+            if (cachedKeys == null) {
+                byte[] saltedPassword = hi(decodeUTF8(password), decodeBase64(salt), iterationCount);
+                byte[] clientKey = hmac(saltedPassword, "Client Key");
+                byte[] serverKey = hmac(saltedPassword, "Server Key");
+                cachedKeys = new CachedKeys(clientKey, serverKey);
+                KEY_CACHE.put(cacheKey, new CachedKeys(clientKey, serverKey));
+            }
+            serverSignature = hmac(cachedKeys.serverKey, authMessage);
+
+            byte[] storedKey = h(cachedKeys.clientKey);
+            byte[] clientSignature = hmac(storedKey, authMessage);
+            byte[] clientProof = xor(cachedKeys.clientKey, clientSignature);
+            return encodeBase64(clientProof);
+        }
+
+        private byte[] decodeBase64(final String str) {
+            return Base64.decode(str);
+        }
+
+        private byte[] decodeUTF8(final String str) throws SaslException {
+            try {
+                return str.getBytes("UTF-8");
+            } catch (UnsupportedEncodingException e) {
+                throw new SaslException("UTF-8 is not a supported encoding.", e);
+            }
+        }
+
+        private String encodeBase64(final String str) throws SaslException {
+            return Base64.encode(decodeUTF8(str));
+        }
+
+        private String encodeBase64(final byte[] bytes) {
+            return Base64.encode(bytes);
+        }
+
+        private String encodeUTF8(final byte[] bytes) throws SaslException {
+            try {
+                return new String(bytes, "UTF-8");
+            } catch (UnsupportedEncodingException e) {
+                throw new SaslException("UTF-8 is not a supported encoding.", e);
+            }
+        }
+
+        private byte[] h(final byte[] data) throws SaslException {
+            try {
+                return MessageDigest.getInstance(hAlgorithm).digest(data);
+            } catch (NoSuchAlgorithmException e) {
+                throw new SaslException(format("Algorithm for '%s' could not be found.", hAlgorithm), e);
+            }
+        }
+
+        private byte[] hi(final byte[] password, final byte[] salt, final int iterations) throws SaslException {
```

Comment:
This algorithm looks a bit different that then one in com.sun.crypto.provider.PBKDF2KeyImpl#deriveKey.  And when I tried reverting this to the original implementation (with appropriate use of PBKDF2WithHmacSHA1 or PBKDF2WithHmacSHA256 based on the mechanism), the SHA-256 spec tests started to fail.  I'm not sure why.

I assume you did this to avoid a Java 8 dependency?

Reply:
Seemed easier to not take a java 8 dependency for trivial code.

Reply:
I agree about the dependency, but I'd like to understand why the tests fail when this is implemented with PBKDF2WithHmacSHA256.  Can you double-check it?

Reply:
I works for me when using PBKDF2WithHmacSHA256 - did you set the byteSize to 32  ?